### DISCUSSION TITLE
How to deal with an stochastic logdensity_fn?

### DISCUSSION THREAD

**Discussion by oarriaga** [blackjax-devs/blackjax#491] [Q&A] [answered]
Hello there!

Often one would like to sample inside a model. However, sampling requires to split a `jax.random.PRNGKey`. This should not be a problem, but it seems one might need to refactor blackjax kernels to pass a `key` to the given `logdensity_fn` to do so.

I had a look at numpyro `seed` [handler](https://num.pyro.ai/en/stable/handlers.html#seed), and they seem to create a changing global key in jax by wrapping it inside a [class field](https://github.com/pyro-ppl/numpyro/blob/7789f39bf4fe510f8f9cc889a9497959e9372f13/numpyro/handlers.py#L717)! I don't know how well this plays in general with `jax.jit`. Has anyone done something similar for creating stochastic log density functions?

**Comment by rlouf** [comment, marked as answer]
Have you tried using the lower-level kernel e.g. [NUTS](https://github.com/blackjax-devs/blackjax/blob/2f19d12d428531f6b507530c7a4c9b0ea7ee401d/blackjax/mcmc/nuts.py#L76) and at each step do something like:

```python
density_key, sample_key = jax.random.split(rng_key)
logdensity_fn = ft.partial(stochastic_logdensity, density_key)
new_state, info = kernel(sample_key, logdensity_fn, step_size, inverse_mass_matrix)
```

You will need to do that for adaptation as well, and can take inspiration from the window adaptation's [implementation](https://github.com/blackjax-devs/blackjax/blob/2f19d12d428531f6b507530c7a4c9b0ea7ee401d/blackjax/kernels.py#L738). This is slightly more cumbersome than the high-level interface but should work fine.
